File: packages/snow_draw_core/lib/draw/store/middleware/middlewares/history_middleware.dart
import '../../../actions/draw_actions.dart';
import '../../../actions/history_policy.dart';
import '../../../elements/types/serial_number/serial_number_dependencies.dart';
import '../../../history/history_metadata.dart';
import '../../../history/recordable.dart';
import '../../../models/draw_state.dart';
import '../../../models/element_state.dart';
import '../../../models/interaction_state.dart';
import '../../history_change_set.dart';
import '../../snapshot.dart';
import '../history_recording_error.dart';
import '../middleware_base.dart';
import '../middleware_context.dart';

/// History middleware that manages undo/redo snapshots.
///
/// It handles:
/// - Recording snapshots for recordable actions
/// - Undo/redo operations
/// - History clearing
/// - Batch mode awareness
///
/// Note: This currently uses delta-based history.
/// Future enhancement: Migrate to Event Sourcing for even better performance.
class HistoryMiddleware extends MiddlewareBase {
  const HistoryMiddleware();

  @override
  String get name => 'History';

  @override
  int get priority => 400;

  @override
  bool shouldExecute(DispatchContext context) {
    final action = context.action;
    final log = context.drawContext.log.history;

    if (action is Undo || action is Redo || action is ClearHistory) {
      log.trace('History middleware executing', {
        'action': action.runtimeType.toString(),
        'traceId': context.traceId,
      });
      return true;
    }

    final policy = _resolveHistoryPolicy(context, action);
    if (policy != HistoryPolicy.record) {
      log.trace('History middleware skipped', {
        'action': action.runtimeType.toString(),
        'reason': 'policy',
        'policy': policy.name,
      });
      return false;
    }

    if (context.isBatching) {
      log.trace('History middleware skipped', {
        'action': action.runtimeType.toString(),
        'reason': 'batching',
      });
      return false;
    }

    return true;
  }

  @override
  Future<DispatchContext> invoke(
    DispatchContext context,
    NextFunction next,
  ) async {
    final action = context.action;
    switch (action) {
      case Undo _:
        return _handleUndo(context, next);
      case Redo _:
        return _handleRedo(context, next);
      case ClearHistory _:
        context.drawContext.log.history.trace('History clear requested', {
          'traceId': context.traceId,
        });
        context.historyManager.clear();
        return next(context);
      default:
        final updatedContext = await next(context);
        _recordHistory(updatedContext, action);
        return updatedContext;
    }
  }

  Future<DispatchContext> _handleUndo(
    DispatchContext context,
    NextFunction next,
  ) => _handleHistoryNavigation(
    context: context,
    next: next,
    actionName: 'undo',
    canNavigate: context.historyManager.canUndo,
    restore: context.historyManager.undo,
  );

  Future<DispatchContext> _handleRedo(
    DispatchContext context,
    NextFunction next,
  ) => _handleHistoryNavigation(
    context: context,
    next: next,
    actionName: 'redo',
    canNavigate: context.historyManager.canRedo,
    restore: context.historyManager.redo,
  );

  Future<DispatchContext> _handleHistoryNavigation({
    required DispatchContext context,
    required NextFunction next,
    required String actionName,
    required bool canNavigate,
    required DrawState? Function(DrawState currentState) restore,
  }) {
    context.drawContext.log.history.trace('History $actionName requested', {
      'traceId': context.traceId,
    });
    if (!canNavigate) {
      return next(context);
    }

    final restoredState = restore(context.currentState);
    if (restoredState == null) {
      return next(context);
    }

    return next(context.withCurrentState(restoredState));
  }

  void _recordHistory(DispatchContext context, DrawAction action) {
    final log = context.drawContext.log.history;
    final metadata = _buildMetadata(context, action);
    final changes = _buildChangeSet(context, action, metadata);
    final includeSelection = context.includeSelectionInHistory;
    final coalescing = action.historyCoalescing;

    try {
      final useIncremental =
          changes != null &&
          !_requiresPersistentSnapshots(action: action, changes: changes);

      final snapshotBefore = _buildSnapshotBefore(
        context: context,
        action: action,
        changes: changes,
        includeSelection: includeSelection,
        useIncremental: useIncremental,
      );
      final snapshotAfter = _buildSnapshotAfter(
        context: context,
        changes: changes,
        includeSelection: includeSelection,
        useIncremental: useIncremental,
      );

      final recorded = context.historyManager.record(
        snapshotBefore,
        snapshotAfter,
        metadata: metadata,
        changes: changes,
        coalescing: coalescing,
        currentState: context.initialState,
        nextState: context.currentState,
      );
      log.trace('History record evaluated', {
        'action': action.runtimeType.toString(),
        'recorded': recorded,
        'description': metadata?.description,
        'coalescingKey': coalescing?.key,
        'traceId': context.traceId,
      });
    } on Object catch (error, stackTrace) {
      Error.throwWithStackTrace(
        HistoryRecordingError(
          action: action.runtimeType.toString(),
          cause: error,
        ),
        stackTrace,
      );
    }
  }

  HistorySnapshot _buildSnapshotBefore({
    required DispatchContext context,
    required DrawAction action,
    required HistoryChangeSet? changes,
    required bool includeSelection,
    required bool useIncremental,
  }) {
    if (action.requiresPreActionSnapshot) {
      if (useIncremental) {
        final resolvedChanges = changes!;
        return context.snapshotBuilder.buildIncrementalSnapshotBeforeAction(
          currentState: context.initialState,
          action: action,
          changes: resolvedChanges,
          includeSelection: includeSelection,
        );
      }
      return context.snapshotBuilder.buildSnapshotBeforeAction(
        currentState: context.initialState,
        action: action,
        includeSelection: includeSelection,
      );
    }

    return _buildSnapshotFromState(
      context: context,
      state: context.initialState,
      changes: changes,
      includeSelection: includeSelection,
      useIncremental: useIncremental,
    );
  }

  HistorySnapshot _buildSnapshotAfter({
    required DispatchContext context,
    required HistoryChangeSet? changes,
    required bool includeSelection,
    required bool useIncremental,
  }) => _buildSnapshotFromState(
    context: context,
    state: context.currentState,
    changes: changes,
    includeSelection: includeSelection,
    useIncremental: useIncremental,
  );

  HistorySnapshot _buildSnapshotFromState({
    required DispatchContext context,
    required DrawState state,
    required HistoryChangeSet? changes,
    required bool includeSelection,
    required bool useIncremental,
  }) {
    if (useIncremental) {
      return context.snapshotBuilder.buildIncrementalSnapshotFromState(
        state: state,
        changes: changes!,
        includeSelection: includeSelection,
      );
    }

    return PersistentSnapshot.fromState(
      state,
      includeSelection: includeSelection,
    );
  }

  HistoryPolicy _resolveHistoryPolicy(
    DispatchContext context,
    DrawAction action,
  ) {
    if (action is FinishEdit && _metadataFromEdit(context) == null) {
      return HistoryPolicy.none;
    }
    return switch (action) {
      Recordable _ => HistoryPolicy.record,
      NonRecordable _ => HistoryPolicy.none,
      _ => action.historyPolicy,
    };
  }

  HistoryMetadata? _buildMetadata(DispatchContext context, DrawAction action) =>
      switch (action) {
        final FinishTextEdit finishTextEdit => _metadataFromFinishTextEdit(
          context,
          finishTextEdit,
        ),
        final FinishEdit finishEdit =>
          finishEdit.metadata ?? _metadataFromEdit(context),
        final Recordable recordable => HistoryMetadata(
          description: recordable.historyDescription,
          recordType: recordable.recordType,
        ),
        _ => null,
      };

  HistoryMetadata? _metadataFromFinishTextEdit(
    DispatchContext context,
    FinishTextEdit action,
  ) {
    final outcome = _resolveFinishTextEditOutcome(context, action);
    return switch (outcome.kind) {
      _FinishTextEditOutcomeKind.delete => HistoryMetadata(
        description: 'Delete text',
        recordType: HistoryRecordType.delete,
      ),
      _FinishTextEditOutcomeKind.create => HistoryMetadata(
        description: 'Create text',
        recordType: HistoryRecordType.create,
      ),
      _FinishTextEditOutcomeKind.edit => HistoryMetadata(
        description: 'Edit text',
        recordType: HistoryRecordType.edit,
      ),
      _FinishTextEditOutcomeKind.noop => null,
    };
  }

  ({String elementId, bool isNew, String text}) _resolveFinishTextEditPayload(
    DispatchContext context,
    FinishTextEdit action,
  ) {
    final interaction = context.initialState.application.interaction;
    if (interaction is TextEditingState) {
      return (
        elementId: interaction.elementId,
        isNew: interaction.isNew,
        text: action.text,
      );
    }

    return (
      elementId: action.elementId,
      isNew: action.isNew,
      text: action.text,
    );
  }

  ({String elementId, _FinishTextEditOutcomeKind kind})
  _resolveFinishTextEditOutcome(
    DispatchContext context,
    FinishTextEdit action,
  ) {
    final payload = _resolveFinishTextEditPayload(context, action);
    final hasText = payload.text.trim().isNotEmpty;
    final kind = switch ((payload.isNew, hasText)) {
      (true, false) => _FinishTextEditOutcomeKind.noop,
      (false, false) => _FinishTextEditOutcomeKind.delete,
      (true, true) => _FinishTextEditOutcomeKind.create,
      (false, true) => _FinishTextEditOutcomeKind.edit,
    };
    return (elementId: payload.elementId, kind: kind);
  }

  HistoryMetadata? _metadataFromEdit(DispatchContext context) {
    final interaction = context.initialState.application.interaction;
    if (interaction is! EditingState) {
      return null;
    }

    final operation = context.drawContext.editOperations.getOperation(
      interaction.operationId,
    );
    if (operation == null || !operation.recordsHistory) {
      return null;
    }

    return operation.createHistoryMetadata(
      context: interaction.context,
      transform: interaction.currentTransform,
    );
  }

  HistoryChangeSet? _buildChangeSet(
    DispatchContext context,
    DrawAction action,
    HistoryMetadata? metadata,
  ) {
    final selectionChanged = _selectionChanged(context);
    final actionChangeSet = switch (action) {
      FinishEdit _ => _buildFinishEditChangeSet(
        context: context,
        metadata: metadata,
        selectionChanged: selectionChanged,
      ),
      final ChangeElementZIndex zIndex => _buildZIndexChangeSet(
        context: context,
        elementIds: {zIndex.elementId},
        selectionChanged: selectionChanged,
      ),
      final ChangeElementsZIndex zIndices => _buildZIndexChangeSet(
        context: context,
        elementIds: zIndices.elementIds.toSet(),
        selectionChanged: selectionChanged,
      ),
      final DeleteElements delete => _buildDeleteElementsChangeSet(
        context: context,
        action: delete,
        selectionChanged: selectionChanged,
      ),
      FinishCreateElement _ => _buildFinishCreateElementChangeSet(
        context: context,
        selectionChanged: selectionChanged,
      ),
      final FinishTextEdit finishTextEdit => _buildFinishTextEditChangeSet(
        context: context,
        action: finishTextEdit,
        selectionChanged: selectionChanged,
      ),
      UpdateGlobalElements _ => HistoryChangeSet(
        globalElementsChanged: true,
        selectionChanged: selectionChanged,
      ),
      DuplicateElements _ => _buildDuplicateElementsChangeSet(
        context: context,
        selectionChanged: selectionChanged,
      ),
      _ => null,
    };

    return actionChangeSet ??
        _buildMetadataChangeSet(
          metadata: metadata,
          selectionChanged: selectionChanged,
        );
  }

  bool _selectionChanged(DispatchContext context) =>
      context.includeSelectionInHistory &&
      context.initialState.domain.selection !=
          context.currentState.domain.selection;

  HistoryChangeSet? _buildFinishEditChangeSet({
    required DispatchContext context,
    required HistoryMetadata? metadata,
    required bool selectionChanged,
  }) {
    final affected = metadata?.affectedElementIds ?? const <String>{};
    if (affected.isEmpty) {
      return null;
    }

    final expandedAffectedIds = _expandModifiedIdsForArrowBindings(
      elements: context.initialState.domain.document.elements,
      modifiedIds: affected,
    );
    return HistoryChangeSet(
      modifiedIds: expandedAffectedIds,
      selectionChanged: selectionChanged,
    );
  }

  HistoryChangeSet _buildZIndexChangeSet({
    required DispatchContext context,
    required Set<String> elementIds,
    required bool selectionChanged,
  }) {
    final reordered = _didElementOrderChange(context);
    return HistoryChangeSet(
      modifiedIds: elementIds,
      orderChanged: true,
      selectionChanged: selectionChanged,
      reindexZIndices: reordered,
    );
  }

  HistoryChangeSet? _buildDeleteElementsChangeSet({
    required DispatchContext context,
    required DeleteElements action,
    required bool selectionChanged,
  }) {
    final document = context.initialState.domain.document;
    final beforeElements = context.initialState.domain.document.elements;
    final removedIds = expandSerialNumberBoundTextIds(
      elements: beforeElements,
      seedIds: action.elementIds.where(document.elementMap.containsKey),
    );
    if (removedIds.isEmpty) {
      return selectionChanged ? HistoryChangeSet(selectionChanged: true) : null;
    }
    final modifiedIds = collectDependentElementIds(
      elements: beforeElements,
      targetIds: removedIds,
      excludedIds: removedIds,
    );
    return HistoryChangeSet(
      modifiedIds: modifiedIds,
      removedIds: removedIds,
      orderChanged: true,
      selectionChanged: selectionChanged,
    );
  }

  HistoryChangeSet? _buildFinishCreateElementChangeSet({
    required DispatchContext context,
    required bool selectionChanged,
  }) {
    final interaction = context.initialState.application.interaction;
    if (interaction is! CreatingState) {
      return null;
    }
    return HistoryChangeSet(
      addedIds: {interaction.elementId},
      orderChanged: true,
      selectionChanged: selectionChanged,
    );
  }

  HistoryChangeSet? _buildFinishTextEditChangeSet({
    required DispatchContext context,
    required FinishTextEdit action,
    required bool selectionChanged,
  }) {
    final outcome = _resolveFinishTextEditOutcome(context, action);
    return switch (outcome.kind) {
      _FinishTextEditOutcomeKind.delete => HistoryChangeSet(
        modifiedIds: _dependentIdsBoundToDeletedText(
          elements: context.initialState.domain.document.elements,
          textElementId: outcome.elementId,
        ),
        removedIds: {outcome.elementId},
        orderChanged: true,
        selectionChanged: selectionChanged,
      ),
      _FinishTextEditOutcomeKind.create => HistoryChangeSet(
        addedIds: {outcome.elementId},
        orderChanged: true,
        selectionChanged: selectionChanged,
      ),
      _FinishTextEditOutcomeKind.edit => HistoryChangeSet(
        modifiedIds: {outcome.elementId},
        selectionChanged: selectionChanged,
      ),
      _FinishTextEditOutcomeKind.noop => null,
    };
  }

  HistoryChangeSet? _buildDuplicateElementsChangeSet({
    required DispatchContext context,
    required bool selectionChanged,
  }) {
    final addedIds = _addedElementIds(
      before: context.initialState.domain.document.elements,
      after: context.currentState.domain.document.elements,
    );
    if (addedIds.isEmpty) {
      return null;
    }
    return HistoryChangeSet(
      addedIds: addedIds,
      orderChanged: true,
      selectionChanged: selectionChanged,
    );
  }

  HistoryChangeSet? _buildMetadataChangeSet({
    required HistoryMetadata? metadata,
    required bool selectionChanged,
  }) {
    final affected = metadata?.affectedElementIds ?? const <String>{};
    if (affected.isEmpty) {
      return null;
    }
    return HistoryChangeSet(
      modifiedIds: affected,
      selectionChanged: selectionChanged,
    );
  }

  bool _requiresPersistentSnapshots({
    required DrawAction action,
    required HistoryChangeSet changes,
  }) =>
      (action is ChangeElementZIndex || action is ChangeElementsZIndex) &&
      !changes.reindexZIndices;

  bool _didElementOrderChange(DispatchContext context) {
    final before = context.initialState.domain.document.elements;
    final after = context.currentState.domain.document.elements;
    if (before.length != after.length) {
      return true;
    }
    for (var index = 0; index < before.length; index++) {
      if (before[index].id != after[index].id) {
        return true;
      }
    }
    return false;
  }

  Set<String> _expandModifiedIdsForArrowBindings({
    required Iterable<ElementState> elements,
    required Set<String> modifiedIds,
  }) => {
    ...modifiedIds,
    ...collectDependentElementIds(
      elements: elements,
      targetIds: modifiedIds,
      excludedIds: modifiedIds,
      includeSerialBindings: false,
    ),
  };

  Set<String> _addedElementIds({
    required Iterable<ElementState> before,
    required Iterable<ElementState> after,
  }) {
    final beforeIds = <String>{for (final element in before) element.id};
    return {
      for (final element in after)
        if (!beforeIds.contains(element.id)) element.id,
    };
  }

  Set<String> _dependentIdsBoundToDeletedText({
    required Iterable<ElementState> elements,
    required String textElementId,
  }) => collectDependentElementIds(
    elements: elements,
    targetIds: {textElementId},
    excludedIds: {textElementId},
  );
}

enum _FinishTextEditOutcomeKind { create, edit, delete, noop }

File: packages/snow_draw_rust_engine/src/draw/store/middleware/middlewares/history_middleware.rs
#![allow(dead_code)]

use std::any::Any;
use std::collections::HashSet;
use std::error::Error;
use std::fmt;
use std::panic::{catch_unwind, AssertUnwindSafe};

use crate::draw::actions::draw_actions::{
    ClearHistory, CreateSerialNumberTextElements, DrawAction, FinishEdit, FinishTextEdit, Redo,
    Undo, UpdateGlobalElements,
};
use crate::draw::actions::history_policy::HistoryPolicy;
use crate::draw::history::history_metadata::{HistoryMetadata, HistoryRecordType};
use crate::draw::history::recordable::{HistoryRecordType as ActionHistoryRecordType, Recordable};
use crate::draw::models::interaction_state::InteractionState;
use crate::draw::store::history_manager::HistoryCoalescing as StoreHistoryCoalescing;
use crate::draw::store::middleware::history_recording_error::HistoryRecordingError;
use crate::draw::store::middleware::middleware_context::DispatchContext;
use crate::draw::store::middleware::middleware_pipeline::{
    DispatchFuture, Middleware, MiddlewareError, MiddlewareInvocationResult, NextFunction,
};

/// History middleware that manages undo/redo snapshots.
///
/// It handles:
/// - recording snapshots for recordable actions,
/// - undo/redo operations,
/// - history clearing,
/// - batch-mode awareness.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct HistoryMiddleware;

impl HistoryMiddleware {
    pub const PRIORITY: i32 = 400;
    pub const NAME: &'static str = "History";

    pub const fn new() -> Self {
        Self
    }

    fn invoke_runtime(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        if context.action.as_any().is::<Undo>() {
            return self.handle_undo(context, next);
        }
        if context.action.as_any().is::<Redo>() {
            return self.handle_redo(context, next);
        }
        if context.action.as_any().is::<ClearHistory>() {
            return self.handle_clear_history(context, next);
        }

        self.handle_recordable_action(context, next)
    }

    fn handle_clear_history(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        {
            let mut manager = context
                .history_manager
                .lock()
                .unwrap_or_else(|poisoned| poisoned.into_inner());
            manager.clear();
        }

        Box::pin(async move { next(context).await })
    }

    fn handle_recordable_action(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        let policy = self.resolve_history_policy(&context, context.action.as_ref());
        if policy != HistoryPolicy::Record || context.is_batching {
            return Box::pin(async move { next(context).await });
        }

        let middleware = *self;
        Box::pin(async move {
            let updated_context = next(context).await?;
            if updated_context.has_error() {
                return Ok(updated_context);
            }

            middleware.record_history(&updated_context)?;
            Ok(updated_context)
        })
    }

    fn handle_undo(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        Box::pin(async move {
            let restored_state = {
                let mut manager = context
                    .history_manager
                    .lock()
                    .unwrap_or_else(|poisoned| poisoned.into_inner());
                if !manager.can_undo() {
                    None
                } else {
                    manager.undo(&context.current_state)
                }
            };

            match restored_state {
                Some(state) => next(context.with_current_state(state)).await,
                None => next(context).await,
            }
        })
    }

    fn handle_redo(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        Box::pin(async move {
            let restored_state = {
                let mut manager = context
                    .history_manager
                    .lock()
                    .unwrap_or_else(|poisoned| poisoned.into_inner());
                if !manager.can_redo() {
                    None
                } else {
                    manager.redo(&context.current_state)
                }
            };

            match restored_state {
                Some(state) => next(context.with_current_state(state)).await,
                None => next(context).await,
            }
        })
    }

    fn resolve_history_policy(
        &self,
        context: &DispatchContext,
        action: &dyn DrawAction,
    ) -> HistoryPolicy {
        if let Some(action) = action.as_any().downcast_ref::<FinishEdit>() {
            if self.metadata_from_finish_edit(context, action).is_none() {
                return HistoryPolicy::None;
            }
        }

        action.history_policy()
    }

    fn build_metadata(
        &self,
        context: &DispatchContext,
        action: &dyn DrawAction,
    ) -> Option<HistoryMetadata> {
        if let Some(action) = action.as_any().downcast_ref::<FinishTextEdit>() {
            return self.metadata_from_finish_text_edit(action);
        }
        if let Some(action) = action.as_any().downcast_ref::<FinishEdit>() {
            return self.metadata_from_finish_edit(context, action);
        }
        if let Some(action) = action.as_any().downcast_ref::<UpdateGlobalElements>() {
            return Some(recordable_metadata(action));
        }
        if let Some(action) = action
            .as_any()
            .downcast_ref::<CreateSerialNumberTextElements>()
        {
            return Some(recordable_metadata(action));
        }

        None
    }

    fn metadata_from_finish_edit(
        &self,
        context: &DispatchContext,
        action: &FinishEdit,
    ) -> Option<HistoryMetadata> {
        action
            .metadata
            .clone()
            .or_else(|| self.metadata_from_edit(context))
    }

    fn metadata_from_finish_text_edit(&self, action: &FinishTextEdit) -> Option<HistoryMetadata> {
        let has_text = !action.text.trim().is_empty();
        if !has_text && action.is_new {
            return None;
        }

        let (description, record_type) = match (action.is_new, has_text) {
            (false, false) => ("Delete text", HistoryRecordType::Delete),
            (true, true) => ("Create text", HistoryRecordType::Create),
            (false, true) => ("Edit text", HistoryRecordType::Edit),
            _ => return None,
        };

        Some(HistoryMetadata::new(
            description.to_owned(),
            record_type,
            HashSet::new(),
            None,
            None,
        ))
    }

    fn metadata_from_edit(&self, context: &DispatchContext) -> Option<HistoryMetadata> {
        let InteractionState::Editing(interaction) = &context.initial_state.application.interaction
        else {
            return None;
        };

        let operation = context
            .draw_context
            .edit_operations
            .get_operation(interaction.operation_id)?;
        if !operation.records_history() {
            return None;
        }

        Some(
            operation.create_history_metadata(&interaction.context, &interaction.current_transform),
        )
    }

    fn record_history(&self, context: &DispatchContext) -> Result<(), MiddlewareError> {
        let metadata = self.build_metadata(context, context.action.as_ref());
        let include_selection = context.include_selection_in_history;
        let coalescing = context
            .action
            .history_coalescing()
            .map(|hint| StoreHistoryCoalescing::new(hint.key.clone(), Some(hint.window)));

        let snapshot_before = context
            .snapshot_builder
            .build_snapshot_from_state(&context.initial_state, include_selection);
        let snapshot_after = context
            .snapshot_builder
            .build_snapshot_from_state(&context.current_state, include_selection);

        let record_result = catch_unwind(AssertUnwindSafe(|| {
            let mut manager = context
                .history_manager
                .lock()
                .unwrap_or_else(|poisoned| poisoned.into_inner());

            manager.record(
                snapshot_before,
                snapshot_after,
                metadata.clone(),
                coalescing,
                Some(&context.initial_state),
                None,
            )
        }));

        match record_result {
            Ok(_recorded) => Ok(()),
            Err(payload) => {
                let panic_error =
                    PanicHistoryRecordingError::new(panic_payload_to_string(payload.as_ref()));
                let error =
                    HistoryRecordingError::with_cause(context.action.action_name(), panic_error);
                Err(MiddlewareError::message(error.to_string()))
            }
        }
    }
}

impl Middleware for HistoryMiddleware {
    fn invoke(
        &self,
        context: DispatchContext,
        next: NextFunction,
    ) -> DispatchFuture<MiddlewareInvocationResult> {
        self.invoke_runtime(context, next)
    }

    fn priority(&self) -> i32 {
        Self::PRIORITY
    }

    fn name(&self) -> &str {
        Self::NAME
    }
}

fn recordable_metadata<T>(recordable: &T) -> HistoryMetadata
where
    T: Recordable,
{
    HistoryMetadata::new(
        recordable.history_description(),
        map_record_type(recordable.record_type()),
        HashSet::new(),
        None,
        None,
    )
}

fn map_record_type(value: ActionHistoryRecordType) -> HistoryRecordType {
    match value {
        ActionHistoryRecordType::Edit => HistoryRecordType::Edit,
        ActionHistoryRecordType::Create => HistoryRecordType::Create,
        ActionHistoryRecordType::Delete => HistoryRecordType::Delete,
        ActionHistoryRecordType::Style => HistoryRecordType::Style,
        ActionHistoryRecordType::Selection => HistoryRecordType::Selection,
        ActionHistoryRecordType::Other => HistoryRecordType::Other,
    }
}

#[derive(Debug)]
struct PanicHistoryRecordingError {
    message: String,
}

impl PanicHistoryRecordingError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for PanicHistoryRecordingError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "history recording panicked: {}", self.message)
    }
}

impl Error for PanicHistoryRecordingError {}

fn panic_payload_to_string(payload: &(dyn Any + Send)) -> String {
    if let Some(value) = payload.downcast_ref::<String>() {
        return value.clone();
    }
    if let Some(value) = payload.downcast_ref::<&str>() {
        return (*value).to_owned();
    }
    "panic without message".to_owned()
}
